widget/richtext.go

```diff
@@ -0,0 +1,734 @@
+package widget
+
+import (
+	"image/color"
+	"math"
+	"strings"
+	"sync"
+	"unicode"
+
+	"fyne.io/fyne/v2"
+	"fyne.io/fyne/v2/canvas"
+	"fyne.io/fyne/v2/internal/widget"
+	"fyne.io/fyne/v2/theme"
+)
+
+const (
+	passwordChar = "•"
+)
+
+// RichText represents the base element for a rich text-based widget.
+//
+// Since: 2.1
+type RichText struct {
+	BaseWidget
+	Segments []RichTextSegment
+	Wrapping fyne.TextWrap
+
+	inset     fyne.Size     // this varies due to how the widget works (entry with scroller vs others with padding)
+	rowBounds []rowBoundary // cache for boundaries
+
+	visualCache map[RichTextSegment][]fyne.CanvasObject
+	cacheLock   sync.Mutex
+}
+
+// NewRichText returns a new RichText widget that renders the given text and segments.
+// If no segments are specified it will be converted to a single segment using the default text settings.
+//
+// Since: 2.1
+func NewRichText(segments ...RichTextSegment) *RichText {
+	t := &RichText{Segments: segments}
+	t.updateRowBounds()
+	return t
+}
```

Comment:
If `ExtendBaseWidget` is not called here (I understand why it is not desired here), then `MinSize()` should be reimplemented to call it, as there are sometimes where `MinSize` is called before `CreateRenderer`. In that case min size will be zero because there is no widget implementation.

Reply:
Excellent, thanks for this - that fixes the flicker that @charlesdaniels reported :)

Reply:
Fixed :)